src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1571,32 +960,87 @@ where
     ((ok_streams, err_stream), None)
 }
 
+/// Reclocks an `IntoTime` frontier stream into a `FromTime` frontier stream. This is used for the
+/// feedback edge that converts the `IntoTime` resumption frontier into the `FromTime` frontier
+/// that is used with the source's `OffsetCommiter`.
+fn reclock_resume_upper<FromTime, IntoTime>(
+    mut resume_rx: UnboundedReceiver<EventCore<IntoTime, Vec<()>>>,
+    mut reclock_follower: ReclockFollower<FromTime, IntoTime>,
+    worker_id: usize,
+    id: GlobalId,
+) -> impl futures::stream::Stream<Item = Antichain<FromTime>>
+where
+    FromTime: SourceTimestamp,
+    IntoTime: Timestamp + Lattice + Display,
+{
+    async_stream::stream!({
+        let mut resume_upper = MutableAntichain::new_bottom(Timestamp::minimum());
+        let mut prev_upper = reclock_follower.since().to_owned();
+
+        while let Some(event) = resume_rx.recv().await {
+            if let EventCore::Progress(changes) = event {
+                resume_upper.update_iter(changes);
+
+                let source_upper = loop {
+                    match reclock_follower.source_upper_at_frontier(resume_upper.frontier()) {
+                        Ok(frontier) => break frontier,
+                        Err(ReclockError::BeyondUpper(_) | ReclockError::Uninitialized) => {
+                            tokio::time::sleep(Duration::from_millis(100)).await
+                        }
```

Comment:
```suggestion
                        Err(ReclockError::BeyondUpper(_) | ReclockError::Uninitialized) => {
                            // Note that we can hold back ingestion of the `resume_rx` indefinitely, because it will not produce unbounded progress messages unless the reclock trace is being progressed, which will unblock this loop!
                            tokio::time::sleep(Duration::from_millis(100)).await
                        }
```